src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -633,107 +741,111 @@ public static String Join<T>(String separator, IEnumerable<T> values)
 
-        // Joins an array of strings together as one string with a separator between each original string.
-        //
-        [System.Security.SecuritySafeCritical]  // auto-generated
-        public unsafe static String Join(String separator, String[] value, int startIndex, int count) {
-            //Range check the array
             if (value == null)
+            {
                 throw new ArgumentNullException(nameof(value));
-
+            }
             if (startIndex < 0)
+            {
                 throw new ArgumentOutOfRangeException(nameof(startIndex), Environment.GetResourceString("ArgumentOutOfRange_StartIndex"));
+            }
             if (count < 0)
+            {
                 throw new ArgumentOutOfRangeException(nameof(count), Environment.GetResourceString("ArgumentOutOfRange_NegativeCount"));
-
+            }
             if (startIndex > value.Length - count)
+            {
                 throw new ArgumentOutOfRangeException(nameof(startIndex), Environment.GetResourceString("ArgumentOutOfRange_IndexCountBuffer"));
-            Contract.EndContractBlock();
-
-            //Treat null as empty string.
-            if (separator == null) {
-                separator = String.Empty;
             }
-
-            //If count is 0, that skews a whole bunch of the calculations below, so just special case that.
-            if (count == 0) {
-                return String.Empty;
+            
+            if (count <= 1)
+            {
+                return count == 0 ?
+                    string.Empty :
+                    value[startIndex] ?? string.Empty;
             }
 
-            if (count == 1) {
-                return value[startIndex] ?? String.Empty;
+            long totalSeparatorsLength = (long)(count - 1) * separatorLength;
+            if (totalSeparatorsLength > int.MaxValue)
+            {
+                throw new OutOfMemoryException();
             }
+            int totalLength = (int)totalSeparatorsLength;
 
-            int jointLength = 0;
-            //Figure out the total length of the strings in value
-            int endIndex = startIndex + count - 1;
-            for (int stringToJoinIndex = startIndex; stringToJoinIndex <= endIndex; stringToJoinIndex++) {
-                string currentValue = value[stringToJoinIndex];
-
-                if (currentValue != null) {
-                    jointLength += currentValue.Length;
+            // Calculate the length of the resultant string so we know how much space to allocate.
+            for (int i = startIndex, end = startIndex + count; i < end; i++)
+            {
+                string currentValue = value[i];
+                if (currentValue != null)
+                {
+                    totalLength += currentValue.Length;
+                    if (totalLength < 0) // Check for overflow
+                    {
+                        throw new OutOfMemoryException();
+                    }
```

Comment:
@stephentoub The original version was incorrect in the first place. `jointLength < 0` does not tell us whether anything overflowed b/c the int could have wrapped around to negative and passed 0 again, since we're dealing with a variable amount of strings.